aead/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+    InvalidTagKeySize,
+
+    #[fail(
+        display = "The given ciphertext is larger than the maximum size allowed by this algorithm"
+    )]
+    CiphertextTooLarge,
+    #[fail(
+        display = "The given plaintext is larger than the maximum size allowed by this algorithm"
+    )]
+    PlaintextTooLarge,
+
+    /// The MAC failed to validate. This should be used to indicated a modified
+    /// ciphertext, modified additional data, or key mismatch.
+    #[fail(display = "The MAC failed to validate")]
+    MacFailure,
+    /// There was an error parsing the padding of this message
+    #[fail(display = "The padding was not correct")]
+    PaddingFailure,
+}
+
+/// A trait used to tag ciphertext inputs/outputs for an authenticated block cipher
+pub trait Ciphertext:
+    Clone
+    + Debug
+    + Display
+    + Into<Vec<u8>>
+    + Send
+    + Sized
+    + Sync
+    + for<'bytes> TryFrom<&'bytes [u8], Error = AeadError>
+{
+    /// Retrieve the length of the ciphertext, in bytes
+    fn len(&self) -> usize;
+}
+
+/// An explicit cryptographic nonce, which may be optionally used for
+/// authenticated block ciphers.
+pub trait Nonce:
+    Clone
+    + Debug
+    + Display
+    + Into<Vec<u8>>
+    + Send
+    + Sized
+    + Sync
+    + for<'bytes> TryFrom<&'bytes [u8], Error = AeadError>
+{
+    /// The length of an explicit nonce, in bytes.
+    type Len: ArrayLength<u8> + Unsigned;
+}
+
+/// A standard nonce implementation for AEAD algorithms which do not use
+/// explicit nonces.
+#[cfg(feature = "serde")]
+#[derive(
+    Clone, Copy, Debug, Default, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize,
+)]
+pub struct EmptyNonce;
+#[cfg(not(feature = "serde"))]
+#[derive(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct EmptyNonce;
+
+/// Output an empty-array style string for the EmptyNonce display.
+impl Display for EmptyNonce {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        write!(f, "[]")
+    }
+}
+
+/// Size restrictions for the Emtpy Nonce type
+impl Nonce for EmptyNonce {
+    type Len = U0;
+}
+
+/// An empty nonce can be converted into an empty vector
+impl Into<Vec<u8>> for EmptyNonce {
+    fn into(self) -> Vec<u8> {
+        Vec::default()
+    }
+}
+
+/// Load this nonce from a byte slice (myst be zero-length)
+impl<'bytes> TryFrom<&'bytes [u8]> for EmptyNonce {
+    type Error = AeadError;
+
+    fn try_from(src: &[u8]) -> Result<Self, Self::Error> {
+        if src.len() != 0 {
+            Err(AeadError::InvalidNonceSize)
+        } else {
+            Ok(EmptyNonce)
+        }
+    }
+}
+
+/// A trait which can support an RFC5116-style authenticated encryption scheme.
+///
+/// In order to reduce the opportunity for misuse, the nonce used for
+/// encryption is meant to be managed internally by the implementation itself.
+pub trait AuthenticatedBlockCipher: Sized + Send + Sync
+where
+    for<'ciphertext> Vec<u8>: From<&'ciphertext Self::Ciphertext>,
```

Comment:
I'm seeing a lot of HRTBs in this PR and having used them in the past I think they're often unnecessary and desirable to avoid if possible. Just a thought.